lib/LanguageServerRename/Adapter/Worse/MemberRenamer.php

```diff
@@ -0,0 +1,127 @@
+use Phpactor\TextDocument\ByteOffset;
+use Phpactor\TextDocument\ByteOffsetRange;
+use Phpactor\TextDocument\TextDocument;
+use Phpactor\TextDocument\TextDocumentLocator;
+use Phpactor\TextDocument\TextEdit as PhpactorTextEdit;
+
+class MemberRenamer implements Renamer
+{
+    /**
+     * @var ReferenceFinder
+     */
+    private $referenceFinder;
+
+    /**
+     * @var TextDocumentLocator
+     */
+    private $locator;
+
+    /**
+     * @var Parser
+     */
+    private $parser;
+
+    public function __construct(
+        ReferenceFinder $referenceFinder,
+        TextDocumentLocator $locator,
+        Parser $parser
+    ) {
+        $this->referenceFinder = $referenceFinder;
+        $this->locator = $locator;
+        $this->parser = $parser;
+    }
+
+    public function getRenameRange(TextDocument $textDocument, ByteOffset $offset): ?ByteOffsetRange
+    {
+        $node = $this->parser->parseSourceFile($textDocument->__toString())->getDescendantNodeAtPosition($offset->toInt());
+
+        if ($node instanceof MethodDeclaration) {
+            return ByteOffsetRange::fromInts($node->name->start, $node->name->getEndPosition());
+        }
+
+        if ($node instanceof Variable && $node->getFirstAncestor(PropertyDeclaration::class)) {
+            return $this->offsetRangeFromToken($node->name, true);
+        }
+
+        if (
+            $node instanceof Variable &&
+            (
+                $node->getFirstAncestor(ScopedPropertyAccessExpression::class) ||
+                $node->getFirstAncestor(MemberAccessExpression::class)
+            )
+        ) {
+            return $this->offsetRangeFromToken($node->name, true);
+        }
+
+        if ($node instanceof MemberAccessExpression || $node instanceof ScopedPropertyAccessExpression) {
+            return $this->offsetRangeFromToken($node->memberName, false);
+        }
+
+        if ($node instanceof ConstElement) {
+            return ByteOffsetRange::fromInts($node->name->start, $node->name->getEndPosition());
+        }
+
+        return null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function rename(TextDocument $textDocument, ByteOffset $offset, string $newName): Generator
+    {
+        foreach ($this->referenceFinder->findReferences($textDocument, $offset) as $reference) {
+            if (!$reference->isSurely()) {
+                continue;
+            }
+
+            $textDocument = $this->locator->get($reference->location()->uri());
+            $range = $this->getRenameRange($textDocument, $reference->location()->offset());
+
+            if (null === $range) {
+                continue;
+            }
+
+            yield new LocatedTextEdit(
+                $reference->location()->uri(),
+                PhpactorTextEdit::create(
+                    $range->start(),
+                    $range->end()->toInt() - $range->start()->toInt(),
+                    $newName
+                )
+            );
+        }
+    }
```

Comment:
note that this method is not specific to members. We could probably change the variable renamer to use exactly this logic (e.g. one solution would be having an abstract "reference finder renamer" class). Class renaming would also be very similar.

Reply:
They are all very, very similar, but not quite. We can probably work out a way to group them though.

Reply:
I think the variable renamer could use the same logic if the variable definition locator behaved as expected (e.g. `<>foreach ($foo` is located instead of `foreach (<>$foo`.

Renaming classes is a bit more involved as it requires to change the use statements.

Reply:
I am really for the proper location of references. It gave me great headache.